batch/batch/worker/worker.py

```diff
@@ -138,6 +141,78 @@ def free(self, port):
         self.ports.put_nowait(port)
 
 
+class NetworkNamespace:
+    def __init__(self, subnet_index):
+        self.network_name = uuid.uuid4().hex[:5]
+        self.host_ip = f'10.0.{subnet_index}.10'
+        self.job_ip = f'10.0.{subnet_index}.11'
+        self.port = None
+        self.host_port = None
+
+    async def init(self):
+        veth_host = self.network_name + '-host'
+        veth_job = self.network_name + '-job'
+        await check_shell(
+            f'''
+ip netns add {self.network_name} && \
+    ip link add name {veth_host} type veth peer name {veth_job} && \
+    ip link set dev {veth_host} up && \
+    ip link set {veth_job} netns {self.network_name} && \
+    ip address add {self.host_ip}/24 dev {veth_host}
+    ip -n {self.network_name} link set dev {veth_job} up && \
+    ip -n {self.network_name} link set dev lo up && \
+    ip -n {self.network_name} address add {self.job_ip}/24 dev {veth_job} && \
+    ip -n {self.network_name} route add default via {self.host_ip}'''
+        )
+
+        await check_shell(
+            f'''
+iptables -w 10 --append FORWARD --in-interface {veth_host} --jump ACCEPT && \
+    iptables -w 10 --append FORWARD --out-interface {veth_host} --jump ACCEPT && \
+    iptables -w 10 --table nat --append POSTROUTING --source {self.job_ip}/24 --jump MASQUERADE'''
+        )
+
+        # The default resolver in `/etc/resolv.conf` is a local address (127.0.0.53) which is
+        # inaccessible in other net namespaces since they have their own loop device. So we use
+        # google's public resolver.
+        os.makedirs(f'/etc/netns/{self.network_name}')
+        await check_shell(f'echo nameserver 8.8.8.8 | tee /etc/netns/{self.network_name}/resolv.conf')
+
+    async def expose_port(self, port, host_port):
+        self.port = port
+        self.host_port = host_port
+        await check_shell(
+            f'iptables --table nat --append PREROUTING \\ '
+            f'--match addrtype --dst-type LOCAL \\ '
+            f'--protocol tcp \\ '
+            f'--match tcp --dport {host_port} \\ '
+            f'--jump DNAT --to-destination {self.job_ip}:{port}'
+        )
+
+    async def cleanup(self):
+        # TODO To be tidy we should delete the IP tables rule, though I'm unsure why
+        # it would necessarily be bad to keep there
+        pass
+
+
+class NetworkAllocator:
+    def __init__(self):
+        self.netns = asyncio.Queue()
+
+    async def reserve(self, netns_pool_min_size: int = 64):
+        for subnet_index in range(10, netns_pool_min_size):
+            netns = NetworkNamespace(subnet_index)
```

Comment:
I'm curious to know what worker start-up time is in this PR.

Reply:
Using a bridge might also be the way to go here. The `iptables` rules apply to just the bridge, we give it an IP address range, etc. New namespaces would just add one end of their veth-pair to the bridge. But we would have to lock it down so they can't talk to each other..